rust/src/live.rs

```diff
@@ -482,6 +484,42 @@ pub(crate) fn transaction_apply_live(
     Ok(())
 }
 
+pub(crate) fn applylive_client_finish() -> CxxResult<()> {
+    let cancellable = gio::NONE_CANCELLABLE;
+    let sysroot = &ostree::Sysroot::new_default();
+    sysroot.load(cancellable)?;
+    let repo = &sysroot.get_repo(cancellable)?;
+    let booted = &require_booted_deployment(sysroot)?;
+    let booted_commit = booted.get_csum().expect("csum");
+    let booted_commit = booted_commit.as_str();
+
+    let live_state = get_live_state(booted)?
+        .ok_or_else(|| anyhow!("Failed to find expected apply-live state"))?;
+
+    // It might happen that the live target commit was GC'd somehow; we're not writing
+    // an explicit ref for it.  In that case skip the diff.
+    if !repo.has_object(
+        ostree::ObjectType::Commit,
+        live_state.commit.as_str(),
+        cancellable,
+    )? {
+        return Ok(());
+    }
```

Comment:
I know this is mostly a translation from the C side, though I'm confused how this could ever happen if we've just applied that commit. Feels like we can just drop this check entirely?

Reply:
One problem is we're not writing a ref for it right now.  We probably should, it's just ugly because that's *persistent* state.  But OTOH we could reconcile on startup.  OTOH if we drive livefs support directly into ostree then it'd know to handle this.

I'd prefer to keep this as is based on that.

Reply:
IOW a bug that exists today is if you use `ostree admin cleanup` outside of rpm-ostree.  Which...well, it is tempting to make `cliwrap: true` *also* intercept ostree...

Reply:
I think I added this check originally on the C side before I changed the rpm-ostree GC code into the set of GC roots TBH.

Reply:
Sorry, I'm confused. Isn't this code running at `rpm-ostree ex apply-live` time? I.e. this is called from `rpmostree_ex_builtin_apply_live`, right? So this is catching the race where the commit stuck around long enough for the apply live to work, but somehow not long enough to retrieve the package list from it to print the diff?

Reply:
(Fine as is too, just wasn't clear if we were talking about the same thing.)

Reply:
Since the daemon releases the sysroot lock as soon as the txn finishes, I think this can race today:

cgroup A: rpm-ostree ex apply-live
rpmostreed.service: Completes txn, unlocks sysroot, sends reply to client
cgroup B: ostree admin cleanup
cgroup A: receives reply, tries to compute pkgdiff

Now, probably if the live commit is GC'd it would actually break subsequent `apply-live` invocations and there's basically no way out of us writing a ref here.  I will look at that.

Reply:
> Since the daemon releases the sysroot lock as soon as the txn finishes, I think this can race today:

Hmm, but checking that the object is present before computing the diff doesn't really close the TOCTOU race anyway if there's no locking.

Agreed re. writing a ref. Feel free to add the `lgtm` label if you were planning to do that as a follow-up!

Reply:
Yeah, you're right.  OK I added another commit here to delete that code for now, will look at a ref as followup